sbi/src/api/tsm.rs

```diff
@@ -0,0 +1,280 @@
+    // non-converted memory will result in a failure and not touch the memory.
+    unsafe { ecall_send(&msg).map(|_| ()) }
+}
+
+/// Adds a VCPU with the given id to the specified TVM.
+pub fn add_vcpu(vmid: u64, vcpu_id: u64) -> Result<()> {
+    let msg = SbiMessage::Tee(TvmCpuCreate {
+        guest_id: vmid,
+        vcpu_id,
+    });
+    // Safety: Creating a vcpu doesn't touch any memory owned here.
+    unsafe { ecall_send(&msg).map(|_| ()) }
+}
+
+/// Returns the value of the specified `register` for the given vcpu.
+pub fn get_vcpu_reg(vmid: u64, vcpu_id: u64, register: TvmCpuRegister) -> Result<u64> {
+    let msg = SbiMessage::Tee(TvmCpuGetRegister {
+        guest_id: vmid,
+        vcpu_id,
+        register,
+    });
+    // Safety: `TvmCpuGetRegister` doesn't access our memory.
+    unsafe { ecall_send(&msg) }
+}
+
+/// Sets the value of the specified `register` for the given vcpu.
+pub fn set_vcpu_reg(vmid: u64, vcpu_id: u64, register: TvmCpuRegister, value: u64) -> Result<()> {
+    let msg = SbiMessage::Tee(TvmCpuSetRegister {
+        guest_id: vmid,
+        vcpu_id,
+        register,
+        value,
+    });
+    // Safety: `TvmCpuSetRegister` doesn't access our memory.
+    unsafe { ecall_send(&msg).map(|_| ()) }
+}
+
+/// Declares the confidential region of the guest's physical address space.
+pub fn add_confidential_memory_region(vmid: u64, guest_addr: u64, len: u64) -> Result<()> {
+    let msg = SbiMessage::Tee(TvmAddConfidentialMemoryRegion {
+        guest_id: vmid,
+        guest_addr,
+        len,
+    });
+    // Safety: `TvmAddConfidentialMemoryRegion` doesn't access our memory at all.
+    unsafe { ecall_send(&msg).map(|_| ()) }
+}
+
+/// Declares an address range to be used for emulating MMIO accesses in the guest.
+pub fn add_emulated_mmio_region(vmid: u64, guest_addr: u64, len: u64) -> Result<()> {
+    let msg = SbiMessage::Tee(TvmAddEmulatedMmioRegion {
+        guest_id: vmid,
+        guest_addr,
+        len,
+    });
+    // Safety: Doesn't touch this host's memory.
+    unsafe { ecall_send(&msg).map(|_| ()) }
+}
+
+/// Declares a region that will can be filled with share pages for communication between a TVM and
+/// the host(e.g. virito).
+pub fn add_shared_memory_region(vmid: u64, guest_addr: u64, len: u64) -> Result<()> {
+    let msg = SbiMessage::Tee(TvmAddSharedMemoryRegion {
+        guest_id: vmid,
+        guest_addr,
+        len,
+    });
+    // Safety: `TvmAddSharedMemoryRegion` doesn't affect host memory
+    unsafe { ecall_send(&msg).map(|_| ()) }
+}
+
+/// Copies the data from the pages backing `src_data` to the guest and records their measurement for
+/// attestation.  src_data must be aligned to the given page size.
+pub fn add_measured_pages(
+    vmid: u64,
+    src_data: &[u8],
+    dest_addr: u64,
+    page_type: TsmPageType,
+    guest_addr: u64,
+) -> Result<()> {
+    if src_data
```

Comment:
I'm not sure `src_data` actually needs to be aligned here since it isn't the source or destination address of a mapping -- we're just copying from it. (I notice the implementation currently requires it to be page-aligned, but it seems like that could be relaxed.)

Reply:
Relaxed the constraint here. The implementation can return an error if it wants.